controller_manager/src/resource_manager.cpp

```diff
@@ -0,0 +1,196 @@
+      std::unique_ptr<hardware_interface::components::Sensor>(
+        sensor_component_loader_.createUnmanagedInstance(component_info.class_type)));
+    sensor_components_.back()->configure(component_info);
+  }
+
+  template<class HardwareT, class HardwareInterfaceT>
+  void initialize_hardware(
+    const hardware_interface::HardwareInfo & hardware_info,
+    pluginlib::ClassLoader<HardwareInterfaceT> & loader,
+    std::vector<HardwareT> & container)
+  {
+    // hardware_class_type has to match class name in plugin xml description
+    // TODO(karsten1987) extract package from hardware_class_type
+    // e.g.: <package_vendor>/<system_type>
+    auto interface = std::unique_ptr<HardwareInterfaceT>(
+      loader.createUnmanagedInstance(hardware_info.hardware_class_type));
+    HardwareT actuator(std::move(interface));
+    container.emplace_back(std::move(actuator));
+    container.back().configure(hardware_info);
+  }
+
+  void initialize_actuator(const hardware_interface::HardwareInfo & hardware_info)
+  {
+    using HardwareT = hardware_interface::ActuatorHardware;
+    using HardwareInterfaceT = hardware_interface::ActuatorHardwareInterface;
+    initialize_hardware<HardwareT, HardwareInterfaceT>(
+      hardware_info, actuator_loader_, actuators_);
+  }
+
+  void initialize_sensor(const hardware_interface::HardwareInfo & hardware_info)
+  {
+    using HardwareT = hardware_interface::SensorHardware;
+    using HardwareInterfaceT = hardware_interface::SensorHardwareInterface;
+    initialize_hardware<HardwareT, HardwareInterfaceT>(
+      hardware_info, sensor_loader_, sensors_);
+  }
+
+  void initialize_system(const hardware_interface::HardwareInfo & hardware_info)
+  {
+    using HardwareT = hardware_interface::SystemHardware;
+    using HardwareInterfaceT = hardware_interface::SystemHardwareInterface;
+    initialize_hardware<HardwareT, HardwareInterfaceT>(
+      hardware_info, system_loader_, systems_);
+  }
+
+  // components plugins
+  pluginlib::ClassLoader<hardware_interface::components::Joint> joint_component_loader_;
+  pluginlib::ClassLoader<hardware_interface::components::Sensor> sensor_component_loader_;
+
+  std::vector<std::unique_ptr<hardware_interface::components::Joint>> joint_components_;
+  std::vector<std::unique_ptr<hardware_interface::components::Sensor>> sensor_components_;
+
+  // hardware plugins
+  pluginlib::ClassLoader<hardware_interface::ActuatorHardwareInterface> actuator_loader_;
+  pluginlib::ClassLoader<hardware_interface::SensorHardwareInterface> sensor_loader_;
+  pluginlib::ClassLoader<hardware_interface::SystemHardwareInterface> system_loader_;
+
+  std::vector<hardware_interface::ActuatorHardware> actuators_;
+  std::vector<hardware_interface::SensorHardware> sensors_;
+  std::vector<hardware_interface::SystemHardware> systems_;
+};
+
+ResourceManager::ResourceManager()
+: resource_storage_(std::make_unique<ResourceStorage>())
+{}
+
+ResourceManager::~ResourceManager() = default;
```

Comment:
Would it be useful to make any of these `virtual` so folks can inherit from them?

Reply:
it's an internal(private) class and I don't see a need to inherit from it now. We can add this once we know what would inherit from it.